statev2/src/storage/tx/order_book.rs

```diff
@@ -0,0 +1,308 @@
+    }
+
+    /// Get the order associated with the given ID
+    pub fn get_order_info(
+        &self,
+        order_id: &OrderIdentifier,
+    ) -> Result<Option<NetworkOrder>, StorageError> {
+        let key = order_key(order_id);
+        self.inner().read(ORDERS_TABLE, &key)
+    }
+
+    /// Get the orders associated with a given nullifier
+    pub fn get_orders_by_nullifier(
+        &self,
+        nullifier: Nullifier,
+    ) -> Result<Vec<OrderIdentifier>, StorageError> {
+        let key = nullifier_key(nullifier);
+        self.inner().read(ORDERS_TABLE, &key).map(|set| set.unwrap_or_default())
+    }
+
+    // --- Helpers --- //
+
+    /// Get an order and error if it is not present
+    fn get_order_info_or_err(
+        &self,
+        order_id: &OrderIdentifier,
+    ) -> Result<NetworkOrder, StorageError> {
+        self.get_order_info(order_id)?
+            .ok_or(StorageError::NotFound(ERR_ORDER_NOT_FOUND.to_string()))
+    }
+
+    /// Get the priority of a cluster
+    fn get_cluster_priority(&self, cluster_id: &ClusterId) -> Result<u32, StorageError> {
+        self.inner()
+            .read(PRIORITIES_TABLE, cluster_id)
+            .map(|priority| priority.unwrap_or(CLUSTER_DEFAULT_PRIORITY))
+    }
+}
+
+// -----------
+// | Setters |
+// -----------
+
+impl<'db> StateTxn<'db, RW> {
+    // --- Interface --- //
+
+    /// Add an order to the order book
+    pub fn write_order(&self, order: &NetworkOrder) -> Result<(), StorageError> {
+        let key = order_key(&order.id);
+        self.inner().write(ORDERS_TABLE, &key, order)
+    }
+
+    /// Write the priority of an order
+    pub fn write_order_priority(&self, order: &NetworkOrder) -> Result<(), StorageError> {
+        let cluster_priority = self.get_cluster_priority(&order.cluster)?;
+        let order_priority = ORDER_DEFAULT_PRIORITY;
+
+        self.inner().write(
+            PRIORITIES_TABLE,
+            &order.id,
+            &OrderPriority { cluster_priority, order_priority },
+        )
+    }
+
+    /// Attach a validity proof to an order
+    ///
+    /// Assumed that the proof has been successfully verified higher in the
+    /// callstack
+    pub fn attach_validity_proof(
+        &self,
+        order_id: &OrderIdentifier,
+        proof: OrderValidityProofBundle,
+    ) -> Result<(), StorageError> {
+        // Update the nullifier as per the proof
+        let new_nullifier = proof.reblind_proof.statement.original_shares_nullifier;
+        self.update_order_nullifier_set(order_id, new_nullifier)?;
+
+        // Update the order itself
+        let mut order = self.get_order_info_or_err(order_id)?;
+        order.validity_proofs = Some(proof);
+        order.public_share_nullifier = new_nullifier;
+        order.state = NetworkOrderState::Verified;
+
+        self.write_order(&order)
+    }
+
+    /// Cancel an order in the order book
+    pub fn cancel_order(&self, order_id: &OrderIdentifier) -> Result<(), StorageError> {
```

Comment:
should cancelling an order remove it from the nullifier set?

Reply:
yeah good call